// @horizon/client/lib/index.js
'use strict';

require('rxjs/add/observable/of');

require('rxjs/add/observable/from');

require('rxjs/add/operator/catch');

require('rxjs/add/operator/concatMap');

require('rxjs/add/operator/map');

require('rxjs/add/operator/filter');

var _ast = require('./ast');

var _socket = require('./socket');

var _auth = require('./auth');

var _model = require('./model');

var defaultHost = typeof window !== 'undefined' && window.location && '' + window.location.host || 'localhost:8181';
var defaultSecure = typeof window !== 'undefined' && window.location && window.location.protocol === 'https:' || false;

function Horizon() {
  var _ref = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

  var _ref$host = _ref.host;
  var host = _ref$host === undefined ? defaultHost : _ref$host;
  var _ref$secure = _ref.secure;
  var secure = _ref$secure === undefined ? defaultSecure : _ref$secure;
  var _ref$path = _ref.path;
  var path = _ref$path === undefined ? 'horizon' : _ref$path;
  var _ref$lazyWrites = _ref.lazyWrites;
  var lazyWrites = _ref$lazyWrites === undefined ? false : _ref$lazyWrites;
  var _ref$authType = _ref.authType;
  var authType = _ref$authType === undefined ? 'unauthenticated' : _ref$authType;
  var _ref$keepalive = _ref.keepalive;
  var keepalive = _ref$keepalive === undefined ? 60 : _ref$keepalive;
  var _ref$WebSocketCtor = _ref.WebSocketCtor;
  var WebSocketCtor = _ref$WebSocketCtor === undefined ? WebSocket : _ref$WebSocketCtor;

  // If we're in a redirection from OAuth, store the auth token for
  // this user in localStorage.

  var tokenStorage = new _auth.TokenStorage({ authType: authType, path: path });
  tokenStorage.setAuthFromQueryParams();

  var url = 'ws' + (secure ? 's' : '') + '://' + host + '/' + path;
  var socket = new _socket.HorizonSocket({
    url: url,
    handshakeMaker: tokenStorage.handshake.bind(tokenStorage),
    keepalive: keepalive,
    WebSocketCtor: WebSocketCtor
  });

  // Store whatever token we get back from the server when we get a
  // handshake response
  socket.handshake.subscribe({
    next: function next(handshake) {
      if (authType !== 'unauthenticated') {
        tokenStorage.set(handshake.token);
      }
    },
    error: function error(err) {
      if (/JsonWebTokenError|TokenExpiredError/.test(err.message)) {
        console.error('Horizon: clearing token storage since auth failed');
        tokenStorage.remove();
      }
    }
  });

  // This is the object returned by the Horizon function. It's a
  // function so we can construct a collection simply by calling it
  // like horizon('my_collection')
  function horizon(name) {
    return new _ast.Collection(sendRequest, name, lazyWrites);
  }

  horizon.currentUser = function () {
    return new _ast.UserDataTerm(horizon, socket.handshake, socket);
  };

  horizon.disconnect = function () {
    socket.complete();
  };

  // Dummy subscription to force it to connect to the
  // server. Optionally provide an error handling function if the
  // socket experiences an error.
  // Note: Users of the Observable interface shouldn't need this
  horizon.connect = function () {
    var onError = arguments.length <= 0 || arguments[0] === undefined ? function (err) {
      console.error('Received an error: ' + err);
    } : arguments[0];

    socket.subscribe(function () {}, onError);
  };

  // Either subscribe to status updates, or return an observable with
  // the current status and all subsequent status changes.
  horizon.status = subscribeOrObservable(socket.status);

  // Convenience method for finding out when disconnected
  horizon.onDisconnected = subscribeOrObservable(socket.status.filter(function (x) {
    return x.type === 'disconnected';
  }));

  // Convenience method for finding out when ready
  horizon.onReady = subscribeOrObservable(socket.status.filter(function (x) {
    return x.type === 'ready';
  }));

  // Convenience method for finding out when an error occurs
  horizon.onSocketError = subscribeOrObservable(socket.status.filter(function (x) {
    return x.type === 'error';
  }));

  horizon.utensils = {
    sendRequest: sendRequest,
    tokenStorage: tokenStorage,
    handshake: socket.handshake
  };
  Object.freeze(horizon.utensils);

  horizon._authMethods = null;
  horizon._root = 'http' + (secure ? 's' : '') + '://' + host;
  horizon._horizonPath = horizon._root + '/' + path;
  horizon.authEndpoint = _auth.authEndpoint;
  horizon.hasAuthToken = tokenStorage.hasAuthToken.bind(tokenStorage);
  horizon.aggregate = _model.aggregate;
  horizon.model = _model.model;

  return horizon;

  // Sends a horizon protocol request to the server, and pulls the data
  // portion of the response out.
  function sendRequest(type, options) {
    // Both remove and removeAll use the type 'remove' in the protocol
    var normalizedType = type === 'removeAll' ? 'remove' : type;
    return socket.hzRequest({ type: normalizedType, options: options }) // send the raw request
    .takeWhile(function (resp) {
      return resp.state !== 'complete';
    });
  }
}

function subscribeOrObservable(observable) {
  return function () {
    if (arguments.length > 0) {
      return observable.subscribe.apply(observable, arguments);
    } else {
      return observable;
    }
  };
}

Horizon.Socket = _socket.HorizonSocket;
Horizon.clearAuthTokens = _auth.clearAuthTokens;

module.exports = Horizon;
